server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java

```diff
@@ -0,0 +1,223 @@
+
+public class IssueJsonWriter {
+
+  private static final String ACTIONS_EXTRA_FIELD = "actions";
+  private static final String TRANSITIONS_EXTRA_FIELD = "transitions";
+  private static final String ASSIGNEE_NAME_EXTRA_FIELD = "assigneeName";
+  private static final String REPORTER_NAME_EXTRA_FIELD = "reporterName";
+  private static final String ACTION_PLAN_NAME_EXTRA_FIELD = "actionPlanName";
+
+  public static final Set<String> EXTRA_FIELDS = ImmutableSet.of(
+    ACTIONS_EXTRA_FIELD, TRANSITIONS_EXTRA_FIELD, ASSIGNEE_NAME_EXTRA_FIELD, REPORTER_NAME_EXTRA_FIELD, ACTION_PLAN_NAME_EXTRA_FIELD);
+
+  private final I18n i18n;
+  private final Durations durations;
+  private final UserSession userSession;
+  private final UserJsonWriter userWriter;
+  private final IssueActionsWriter actionsWriter;
+
+  public IssueJsonWriter(I18n i18n, Durations durations, UserSession userSession, UserJsonWriter userWriter, IssueActionsWriter actionsWriter) {
+    this.i18n = i18n;
+    this.durations = durations;
+    this.userSession = userSession;
+    this.userWriter = userWriter;
+    this.actionsWriter = actionsWriter;
+  }
+
+  public void write(JsonWriter json, Issue issue, Map<String, User> usersByLogin, Map<String, ComponentDto> componentsByUuid,
+    Map<String, ComponentDto> projectsByComponentUuid, Multimap<String, DefaultIssueComment> commentsByIssues, Map<String, ActionPlan> actionPlanByKeys, List<String> extraFields) {
+    json.beginObject();
+
+    String actionPlanKey = issue.actionPlanKey();
+    ComponentDto file = componentsByUuid.get(issue.componentUuid());
+    ComponentDto project = null, subProject = null;
+    if (file != null) {
+      project = projectsByComponentUuid.get(file.uuid());
+      if (!file.projectUuid().equals(file.moduleUuid())) {
+        subProject = componentsByUuid.get(file.moduleUuid());
+      }
+    }
+    Duration debt = issue.debt();
+    Date updateDate = issue.updateDate();
+
+    json
+      .prop("key", issue.key())
+      .prop("component", file != null ? file.getKey() : null)
+      // Only used for the compatibility with the Issues Java WS Client <= 4.4 used by Eclipse
+      .prop("componentId", file != null ? file.getId() : null)
+      .prop("project", project != null ? project.getKey() : null)
+      .prop("subProject", subProject != null ? subProject.getKey() : null)
+      .prop("rule", issue.ruleKey().toString())
+      .prop("status", issue.status())
+      .prop("resolution", issue.resolution())
+      .prop("severity", issue.severity())
+      .prop("message", issue.message())
+      .prop("line", issue.line())
+      .prop("debt", debt != null ? durations.encode(debt) : null)
+      .prop("reporter", issue.reporter())
+      .prop("author", issue.authorLogin())
+      .prop("actionPlan", actionPlanKey)
+      .prop("creationDate", isoDate(issue.creationDate()))
+      .prop("updateDate", isoDate(updateDate))
+      // TODO Remove as part of Front-end rework on Issue Domain
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)